exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/statistics/StatisticsMergeBatch.java

```diff
@@ -0,0 +1,532 @@
+                  + collector.toErrorString());
+        }
+      }
+    }
+  }
+
+  @Override
+  protected boolean setupNewSchema() throws SchemaChangeException {
+    container.clear();
+    buildKeyList();
+    buildOutputContainer();
+    return true;
+  }
+
+  @Override
+  protected IterOutcome doWork() {
+    int outRecordCount = incoming.getRecordCount();
+    HashMap<String, Long> nonNullRows = new HashMap<>();
+
+    for (VectorWrapper<?> vw : incoming) {
+      ValueVector vv = vw.getValueVector();
+      if (vv.getField().getType().getMinorType() != TypeProtos.MinorType.MAP) {
+        //We only expect Maps in the incoming. STOP, if this is not the case.
+        return IterOutcome.STOP;
+      }
+      MapVector mapVec = (MapVector) vv;
+      Map<String, ValueHolder> statMap = aggregationMap.get(vv.getField().getLastName());
+
+      for (ValueVector mapElt : mapVec) {
+        if (vv.getField().getLastName().equalsIgnoreCase("column")) {
+          BigIntHolder nameHolder;
+          if (statMap.get(mapElt.getField().getLastName()) != null) {
+            nameHolder = (BigIntHolder) statMap.get(mapElt.getField().getLastName());
+          } else {
+            nameHolder = new BigIntHolder();
+            statMap.put(mapElt.getField().getLastName(), nameHolder);
+          }
+          nameHolder.value = 1;
+        } else if (vv.getField().getLastName().equalsIgnoreCase("statcount")
+            || vv.getField().getLastName().equalsIgnoreCase("nonnullstatcount")) {
+          BigIntHolder sumHolder;
+          String colName = mapElt.getField().getLastName();
+          if (statMap.get(colName) != null) {
+            sumHolder = (BigIntHolder) statMap.get(colName);
+          } else {
+            sumHolder = new BigIntHolder();
+            statMap.put(colName, sumHolder);
+          }
+          //TODO: assert size = 1
+          //TODO: logger
+          Object val = mapElt.getAccessor().getObject(0);
+          if (val != null) {
+            sumHolder.value += (long)val;
+          }
+        } else if (vv.getField().getLastName().equalsIgnoreCase("sum_width")) {
+          NullableFloat8Holder sumHolder;
+          String colName = mapElt.getField().getLastName();
+          if (statMap == null) {
+            statMap = aggregationMap.get(functions.get(vv.getField().getLastName()));
+          }
+          if (statMap.get(colName) != null) {
+            sumHolder = (NullableFloat8Holder) statMap.get(colName);
+          } else {
+            sumHolder = new NullableFloat8Holder();
+            statMap.put(colName, sumHolder);
+          }
+          //TODO: assert size = 1
+          //TODO: logger
+          Object val = mapElt.getAccessor().getObject(0);
+          if (val != null) {
+            sumHolder.value += (double) val;
+            sumHolder.isSet = 1;
+          }
+        } else if (vv.getField().getLastName().equalsIgnoreCase("hll")) {
```

Comment:
We are hard coding the behavior for each function. (Also the name.) We are doing this in a big if/then statement. Each block has lots of repeated code. Can we make this more modular?

* At the very least, the body of each if statement should be a method so it is easier to follow.
* Those methods should call other methods to do shared tasks. (Please, no copy & paste!)
* Perhaps the behavior should be factored out into a function definition class to achieve the extensibility that appears to be built into other parts of the code.

Overall, this frankly seems far too complex for what it is doing. Something seems to need a good rethink here...